### PR TITLE
arch/armv7-a/r: unify switch context from software interrupt

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arch/arm: unify switch context from software interrupt
arch/armv7-r: unify switch context from software interrupt
arch/armv7-a: unify switch context from software interrupt

## Impact

N/A

## Testing

complete the ostest on armv7-a platform(UP and SMP)

CONFIG_SMP=y
CONFIG_SMP_NCPUS=2

ci check

### CODE REVIEW

**Comment by masayuki2009** [comment]
@anchao 
Let me confirm the intention of this PR.
I know that Cortex-M uses a software interrupt-based context switch.
So, are you going to apply the same approach to arm/armv7-a/r as well?

**Comment by masayuki2009** [comment]
```
====================================================================================
Configuration/Tool: sama5d2-xult/nsh,CONFIG_ARMV7M_TOOLCHAIN_GNU_EABIL
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Disabling CONFIG_ARMV7A_TOOLCHAIN_GNU_EABIL
  Enabling CONFIG_ARMV7M_TOOLCHAIN_GNU_EABIL
  Building NuttX...
make[1]: *** No rule to make target 'arm_savefpu.S', needed by '.depend'.
make[1]: *** No rule to make target 'arm_copyarmstate.c', needed by '.depend'.
make[1]: Target 'depend' not remade because of errors.
make: *** [tools/Unix.mk:584: pass2dep] Error 2
make: Target 'all' not remade because of errors.
/github/workspace/sources/nuttx/tools/testbuild.sh: line 252: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
  Normalize sama5d2-xult/nsh
```

**Comment by anchao** [comment]
> @anchao Let me confirm the intention of this PR. I know that Cortex-M uses a software interrupt-based context switch. So, are you going to apply the same approach to arm/armv7-a/r as well?

Yes, This PR unify the context switch strategy same as Cortex-M, we plan to optimize the context switch and interrupt handing performance in next stage.

**Comment by masayuki2009** [comment]
@anchao 
I tried this PR with sabre-6quad:netnsh_smp (QEMU) and sabre-6quad:smp (Dev board) and confirmed both works.
However, sabre-6quad:netknsh (QEMU) which is not merged and under testing for ADDRENV=y does not work.

```
[    4.020000] exec_module: Executing /mnt/nfs/bin/init
[    4.020000] up_addrenv_select: addrenv=0x10863ee4 oldenv=0x10868ac8
[    4.020000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.020000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.030000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)
[    4.030000] exec_module: initialize the user heap (heapsize=1048576) 
[    4.030000] up_addrenv_kstackalloc: tcb=0x10866820 stacksize=3072
[    4.030000] up_addrenv_clone: src=0x10863ee4 dest=0x10863f58
[    4.030000] up_addrenv_restore: oldenv=0x10868ac8
[    4.040000] up_addrenv_select: addrenv=0x10863f58 oldenv=0
[    4.040000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.040000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.040000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)
[    4.050000] arm_prefetchabort: Prefetch abort. PC: 80202800 IFAR: 00000000 IFSR: 00000002
[    4.050000] up_assert: Assertion failed at file:armv7-a/arm_prefetchabort.c line: 134 task: /mnt/nfs/bin/init
[    4.050000] arm_registerdump: R0: 00000019 R1: deadbeef R2: 00000000  R3: 802027cc
[    4.050000] arm_registerdump: R4: 00000018 R5: 80202018 R6: 00000001  R7: deadbeef
[    4.050000] arm_registerdump: R8: deadbeef SB: 802027e4 SL: 10818140  FP: 80202018
[    4.050000] arm_registerdump: IP: 00000001 SP: 80001a34 LR: 00000003  PC: 80202800
```

**Comment by anchao** [comment]
>> I tried this PR with sabre-6quad:netnsh_smp (QEMU) and sabre-6quad:smp (Dev board) and confirmed both works.
>> However, sabre-6quad:netknsh (QEMU) which is not merged and under testing for ADDRENV=y does not work.

@masayuki2009 san,

Thanks a lot for your verification! could you please upstream the defconfig ? I think this PR broken the kernel mode

**Comment by masayuki2009** [comment]
>Thanks a lot for your verification! could you please upstream the defconfig ? I think this PR broken the kernel mode

@anchao 
I need more refactoring a lot.
For the moment, you can refer to my branches

https://github.com/masayuki2009/incubator-nuttx/tree/imx6-netknsh
https://github.com/masayuki2009/incubator-nuttx-apps/tree/imx6-netknsh

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Comment by anchao** [comment]
> @anchao I need more refactoring a lot. For the moment, you can refer to my branches
> 
> https://github.com/masayuki2009/incubator-nuttx/tree/imx6-netknsh https://github.com/masayuki2009/incubator-nuttx-apps/tree/imx6-netknsh

@masayuki2009 , 

Could you please share the command line of QEMU? I am not sure which machine for sama5d4-ek board

**Comment by masayuki2009** [comment]
> I think this PR broken the kernel mode

@anchao 
Hmm, the latest patch causes assertion.

```
[    4.330000] load_absmodule: Successfully loaded module /mnt/nfs/bin/init
[    4.330000] binfmt_dumpmodule: Module:
[    4.330000] binfmt_dumpmodule:   entrypt:   0x80001a34
[    4.330000] binfmt_dumpmodule:   mapped:    0 size=0
[    4.330000] binfmt_dumpmodule:   alloc:     0 0 0
[    4.330000] binfmt_dumpmodule:   addrenv:   0x18000000
[    4.330000] binfmt_dumpmodule:   stacksize: 2048
[    4.330000] binfmt_dumpmodule:   unload:    0
[    4.330000] exec_module: Executing /mnt/nfs/bin/init
[    4.340000] up_addrenv_select: addrenv=0x10863f04 oldenv=0x10868ac8
[    4.340000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.340000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.340000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)
[    4.340000] exec_module: initialize the user heap (heapsize=1048576) 
[    4.340000] up_addrenv_kstackalloc: tcb=0x10866820 stacksize=3072
[    4.340000] up_addrenv_clone: src=0x10863f04 dest=0x10863f78
[    4.340000] up_addrenv_restore: oldenv=0x10868ac8
[    4.350000] up_addrenv_select: addrenv=0x10863f78 oldenv=0
[    4.350000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.350000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.350000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)
[    4.360000] up_assert: Assertion failed at file:armv7-a/arm_syscall.c line: 242 task: /mnt/nfs/bin/init
[    4.360000] arm_registerdump: R0: 00000001 R1: 00000000 R2: 00000000  R3: 00000000
[    4.360000] arm_registerdump: R4: 108675b8 R5: 00000000 R6: 00000000  R7: 00000000
[    4.360000] arm_registerdump: R8: 00000000 SB: 00000000 SL: 00000000  FP: 1086755c
[    4.360000] arm_registerdump: IP: 00000001 SP: 10867548 LR: 1081783c  PC: 10819d58
[    4.360000] arm_registerdump: CPSR: 600000d3
```

**Comment by masayuki2009** [comment]
>Could you please share the command line of QEMU? I am not sure which machine for sama5d4-ek board

@anchao 
You need to use sabre-6quad:netknsh instead of sama5d4-ek.

```
$ ~/opensource/QEMU/qemu-5.2/build/arm-softmmu/qemu-system-arm -M sabrelite -m 1024 -smp 4 -kernel ./nuttx -nographic
```
Also, please note that you need to set up an NFS server where applications such as init are stored.

**Comment by anchao** [comment]
> @anchao please resolve merge conflicts

Let me do more tests in kernel mode, mark this PR as draft first

**Comment by masayuki2009** [comment]
@anchao 

Please add `-m 1024` option.

**Comment by anchao** [comment]
> @anchao
> 
> Please add `-m 1024` option.

Thanks, it works fine with `-m 1024`，I will investigate further based on this PR!

**Comment by anchao** [comment]
> [    2.170000] up_assert: Assertion failed at file:armv7-a/arm_addrenv.c line: 181 task: nfsroot

@masayuki2009 san,

I have fix this issue caused by SYS_syscall_return mismatch with new syscall definition:
https://github.com/apache/incubator-nuttx/pull/5506/commits/765e5bc52fdf135dc0b0456c58d094a839519178

please review, thank you!

**Comment by masayuki2009** [comment]
>@masayuki2009 san,
>
>I have fix this issue caused by SYS_syscall_return mismatch with new syscall definition:
>https://github.com/apache/incubator-nuttx/commit/765e5bc52fdf135dc0b0456c58d094a839519178
>
>please review, thank you!

@anchao 
Could you create a separate PR for the above commit?

**Comment by anchao** [comment]
> @anchao Could you create a separate PR for the above commit?

@masayuki2009 ,
Ok, new PR for SYS_syscall_return mismatch issue:
https://github.com/apache/incubator-nuttx/pull/5582

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@masayuki2009 could you try this patch with the latest mainline?

**Comment by masayuki2009** [comment]
@anchao 

I tried this PR with my branch (imx6-netknsh) but the nsh prompt did not appear.
How about your situation?

```
[    4.140000] load_absmodule: Successfully loaded module /mnt/nfs/bin/init
[    4.140000] binfmt_dumpmodule: Module:
[    4.140000] binfmt_dumpmodule:   entrypt:   0x80001a34
[    4.140000] binfmt_dumpmodule:   mapped:    0 size=0
[    4.140000] binfmt_dumpmodule:   alloc:     0 0 0
[    4.140000] binfmt_dumpmodule:   addrenv:   0x18000000
[    4.140000] binfmt_dumpmodule:   stacksize: 2048
[    4.140000] binfmt_dumpmodule:   unload:    0
[    4.140000] exec_module: Executing /mnt/nfs/bin/init
[    4.140000] up_addrenv_select: addrenv=0x10863fc4 oldenv=0x10868ac8
[    4.140000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.140000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.150000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)
[    4.150000] exec_module: Initialize the user heap (heapsize=1048576)
[    4.150000] up_addrenv_kstackalloc: tcb=0x10864960 stacksize=3072
[    4.150000] up_addrenv_clone: src=0x10863fc4 dest=0x10864b38
[    4.150000] up_addrenv_restore: oldenv=0x10868ac8
[    4.160000] up_addrenv_select: addrenv=0x10864b38 oldenv=0
[    4.160000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.160000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.160000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)
```

**Comment by anchao** [comment]
> I tried this PR with my branch (imx6-netknsh) but the nsh prompt did not appear. How about your situation?

@masayuki2009 san,

please rebase your branch since this PR based on syscall mismatch fix:
https://github.com/apache/incubator-nuttx/pull/5582

git log:
```
incubator-nuttx$ git log --oneline -7
ca1cda8fec (HEAD) boards/sama5: update the reserved syscall count
4ea1121d87 arch/arm: unify switch context from software interrupt
4481e7ec88 arch/armv7-r: unify switch context from software interrupt
fc68f677c8 arch/armv7-a: unify switch context from software interrupt
3d5b40e2c9 arch/armv7-a: replace SYS_signal_handler_return hardcode
1415c79ae8 arch/arm: replace SYS_syscall_return hardcode from syscall
0b0cab62cf [DO NOT MERGE] sched: group: Defer to call up_addrenv_destroy() for pthread
```

startup log:
```
incubator-nuttx$ qemu-system-arm -net nic -net user,hostfwd=tcp:127.0.0.1:10023-10.0.2.15:23,hostfwd=tcp:127.0.0.1:10021-10.0.2.15:21,hostfwd=tcp:127.0.0.1:15001-10.0.2.15:5001 -M sabrelite -smp 4 -kernel ./nuttx -nographic -m 1024
     
ABDGHIJKNOPQ
[    0.000000] elf_initialize: Registering ELF
....
[    4.510000] load_absmodule: Successfully loaded module /mnt/nfs/bin/init
[    4.510000] binfmt_dumpmodule: Module:
[    4.510000] binfmt_dumpmodule:   entrypt:   0x80001a2c
[    4.510000] binfmt_dumpmodule:   mapped:    0 size=0
[    4.510000] binfmt_dumpmodule:   alloc:     0 0 0
[    4.510000] binfmt_dumpmodule:   addrenv:   0x18000000
[    4.510000] binfmt_dumpmodule:   stacksize: 2048
[    4.510000] binfmt_dumpmodule:   unload:    0
[    4.510000] exec_module: Executing /mnt/nfs/bin/init
[    4.510000] up_addrenv_select: addrenv=0x10864d44 oldenv=0x10868ac8
[    4.510000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.510000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.510000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)
[    4.520000] exec_module: Initialize the user heap (heapsize=1048576)
[    4.520000] up_addrenv_kstackalloc: tcb=0x10864d70 stacksize=3072
[    4.520000] up_addrenv_clone: src=0x10864d44 dest=0x10864f48
[    4.520000] up_addrenv_restore: oldenv=0x10868ac8
[    4.520000] up_addrenv_select: addrenv=0x10864f48 oldenv=0
[    4.520000] up_addrenv_select: text: set l1 entry (paddr=18000000 vaddr=80000000)
[    4.520000] up_addrenv_select: data: set l1 entry (paddr=18028000 vaddr=80100000)
[    4.520000] up_addrenv_select: heap: set l1 entry (paddr=1802a000 vaddr=80200000)

NuttShell (NSH) NuttX-10.2.0
nsh> 


```

**Comment by masayuki2009** [comment]
>please rebase your branch since this PR based on syscall mismatch fix:

I thought I rebased the branch to the latest master.
Perhaps, something was wrong with my branch.
Now the PR works with my branch, thanks.

File: arch/arm/src/arm/arm_fullcontextrestore.S [hunk 1]
```diff
@@ -60,38 +63,14 @@
...
 
-	/* Now recover r0-r1, pc and cpsr, destroying the stack frame */
+	mov		r1, r0				/* R1: regs */
+	mov		r0, #SYS_restore_context	/* R0: restore context */
+	svc		#SYS_syscall			/* Force synchronous SVCall (or Hard Fault) */
 
-	ldmia	sp!, {r0-r1, r15}^
+	/* This call should not return */
 
+	bx		lr				/* Unnecessary ... will not return */
```
[the remark this answers is not in the document]
Comment (anchao) [reply arch/arm/src/arm/arm_fullcontextrestore.S:74]: I will consider fix this in a new PR

**Pull request merged**